### PR TITLE
Refactor VPC CNI in managed control plane

### LINKED ISSUE #3410: Deprecate `DisableVPCCNI` in `AWSManagedControlPlane`
/kind feature
/area provider/eks
/priority important-longterm
/triage accepted
/help

**Describe the solution you'd like**
With #3374 we are introducing a new struct for settings that relate to the VPC CNI  (a.k.a aws-node). We should add a `Disable` field to this and start deprecation of `DisableVPCCNI`.

Any code that relies on `DisableVPCCNI` should be changed to use a field on the managed control plane scope that can internally get the values from `DisableVPCCNI` or `VPCCNI.Disable`.

We will also probably need to update the admission controller to:

- report that `DisableVPCCNI` is deprecated
- check if the values of `DisableVPCCNI` & `VPCCNI.Disable` are set to something different

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Cluster-api-provider-aws version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by Ankitasw**
**What type of PR is this?**
/kind refactor

**What this PR does / why we need it**:
Refactor the disableVPCCNI on the managed control plane inside VpcCni.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes Part of #2355 
Fixes #3410 

**Special notes for your reviewer**:

**Checklist**:

- [x] squashed commits
- [ ] includes documentation
- [x] adds unit tests
- [ ] adds or updates e2e tests

### CODE REVIEW

**Review by AverageMarcus — COMMENTED** [review]
I'm not sure the conversion is right on this, but I could be wrong as I always get confused with the auto generated vs manual stuff.

I was expecting something that converted from the old `.DisableVPCCNI` to the new `.VpcCni.Disable` and vice-versa but I don't see that here. Am I missing something?

**Comment by Ankitasw** [comment]
> I was expecting something that converted from the old .DisableVPCCNI to the new .VpcCni.Disable and vice-versa but I don't see that here. Am I missing something?

I was having the same doubt, but I am not sure how that conversion would wpork, do you have any idea? @MarcusNoble

File: controlplane/eks/api/v1beta1/conversion.go [hunk 1]
```diff
@@ -31,6 +32,13 @@ func (r *AWSManagedControlPlane) ConvertTo(dstRaw conversion.Hub) error {
 	if err := Convert_v1beta1_AWSManagedControlPlane_To_v1beta2_AWSManagedControlPlane(r, dst, nil); err != nil {
 		return err
 	}
+
```
Comment (AverageMarcus) [inline controlplane/eks/api/v1beta1/conversion.go:35]:
```suggestion
	if r.Spec.DisableVPCCNI != dst.Spec.VpcCni.Disable {
		dst.Spec.VpcCni.Disable = r.Spec.DisableVPCCNI
	}
	
	
```

File: controlplane/eks/api/v1beta1/conversion.go [hunk 2]
```diff
@@ -42,6 +50,9 @@ func (r *AWSManagedControlPlane) ConvertFrom(srcRaw conversion.Hub) error {
 	if err := Convert_v1beta2_AWSManagedControlPlane_To_v1beta1_AWSManagedControlPlane(src, r, nil); err != nil {
 		return err
 	}
+	if err := utilconversion.MarshalData(src, r); err != nil {
```
Comment (AverageMarcus) [inline controlplane/eks/api/v1beta1/conversion.go:53]:
```suggestion

	if r.Spec.DisableVPCCNI != src.Spec.VpcCni.Disable {
		r.Spec.DisableVPCCNI = src.Spec.VpcCni.Disable
	}

	if err := utilconversion.MarshalData(src, r); err != nil {
```

File: controlplane/eks/api/v1beta1/conversion.go (hunk 1 above)
Comment (Ankitasw) [reply controlplane/eks/api/v1beta1/conversion.go:35]: Why if check? What if we directly assign ?

Comment (AverageMarcus) [reply controlplane/eks/api/v1beta1/conversion.go:35]: Good point. No reason to I guess.

File: controlplane/eks/api/v1beta1/conversion.go (hunk 2 above)
Comment (Ankitasw) [reply controlplane/eks/api/v1beta1/conversion.go:53]: Same goes here right? Only I need to add 

> 		r.Spec.DisableVPCCNI = src.Spec.VpcCni.Disable

Comment (AverageMarcus) [reply controlplane/eks/api/v1beta1/conversion.go:53]: Yeah

File: controlplane/eks/api/v1beta1/conversion.go (hunk 1 above)
Comment (Ankitasw) [reply controlplane/eks/api/v1beta1/conversion.go:35]: So this change is already present after `Unmarshal`.

File: controlplane/eks/api/v1beta1/conversion.go (hunk 2 above)
Comment (Ankitasw) [reply controlplane/eks/api/v1beta1/conversion.go:55]: Done

File: controlplane/eks/api/v1beta1/conversion.go (hunk 1 above)
Comment (AverageMarcus) [reply controlplane/eks/api/v1beta1/conversion.go:35]: No, that's not converting from `DisableVPCCNI` => `VpcCni.Disable`

Comment (Ankitasw) [reply controlplane/eks/api/v1beta1/conversion.go:35]: Done

File: test/e2e/suites/unmanaged/unmanaged_CAPI_clusterclass_test.go [hunk 3]
```diff
@@ -66,7 +66,7 @@ var _ = ginkgo.Context("[unmanaged] [Cluster API Framework] [ClusterClass]", fun
 		})
 	})
 
-	ginkgo.Describe("Cluster Upgrade Spec - HA control plane with workers [K8s-Upgrade] [ClusterClass]", func() {
+	ginkgo.FDescribe("Cluster Upgrade Spec - HA control plane with workers [K8s-Upgrade] [ClusterClass]", func() {
```
Comment (Ankitasw) [inline test/e2e/suites/unmanaged/unmanaged_CAPI_clusterclass_test.go:69]: Need to revert this post the test runs

File: controlplane/eks/api/v1beta1/conversion.go (hunk 1 above)
Comment (AverageMarcus) [reply controlplane/eks/api/v1beta1/conversion.go:35]: It still the same. Am I missing something or has GitHub glitched?

Comment (Ankitasw) [reply controlplane/eks/api/v1beta1/conversion.go:35]: Oh i made this change, didn't reflect, I will add again

**Review by AverageMarcus — COMMENTED** [review]
/lgtm

**Comment by richardcase** [comment]
Until the e2e pass:

/hold

**Comment by richardcase** [comment]
/approve

@Ankitasw - looks good, thanks. iI you unhold when the e2e finish and pass

**Comment by Ankitasw** [comment]
The test is flaky for external VPC so retriggered.

**Pull request merged**